code/game/objects/items/science_sampler.dm

```diff
@@ -0,0 +1,334 @@
+/*
+	Sampling
+*/
+/obj/item/sampler
+	name = "science sampler"
+	desc = "A multi-purpose sampling device designed by Zeng-Hu Pharmaceuticals for gathering samples during their research expeditions. It has attachments allowing for sampling of biological tissue, surface soil and water sources."
+	icon = 'icons/obj/item/sampling.dmi'
+	icon_state = "sampler"
+	item_state = "sampler"
+	contained_sprite = TRUE
+	w_class = ITEMSIZE_SMALL
+	var/attachment = SAMPLE_BIO
+	var/obj/item/reagent_containers/glass/beaker/vial/vial
+
+/obj/item/sampler/Initialize(mapload, ...)
+	. = ..()
+	update_icon()
+
+/obj/item/sampler/attackby(obj/item/attacking_item, mob/user, params)
+	if(istype(attacking_item, /obj/item/reagent_containers/glass/beaker/vial))
+		if(!vial)
+			to_chat(user, SPAN_NOTICE("You insert \the [attacking_item] into \the [src]."))
+			vial = attacking_item
+			if(ishuman(user))
+				var/mob/living/carbon/human/H = user
+				H.drop_from_inventory(vial)
+			vial.forceMove(src)
+			update_icon()
+			return TRUE
+		to_chat(user, SPAN_NOTICE("\The [src] already has a vial inserted."))
+	return ..()
+
+/obj/item/sampler/attack_hand(mob/user)
+	if(vial)
+		to_chat(user, SPAN_NOTICE("You remove \the [vial] from \the [src]."))
+		user.put_in_active_hand(vial)
+		vial = null
+		update_icon()
+		return TRUE
+	return ..()
+
+/obj/item/sampler/AltClick(mob/user)
+	attachment = next_in_list(attachment, ALL_SAMPLE_ATTACHMENTS)
+	to_chat(user, SPAN_NOTICE("You switch \the [src] to its [attachment] attachment."))
+	update_icon()
+
+/obj/item/sampler/update_icon()
+	. = ..()
+	ClearOverlays()
+	if(vial)
+		icon_state = "[initial(icon_state)]_loaded"
+		if(vial.reagents.total_volume)
+			AddOverlays(overlay_image(icon, "sampler_full"))
+	else
+		icon_state = initial(icon_state)
+	var/image/I
+	switch(attachment)
+		if(SAMPLE_BIO)
+			I = overlay_image(icon, "bio_attachment")
+		if(SAMPLE_SOIL)
+			I = overlay_image(icon, "soil_attachment")
+		if(SAMPLE_WATER)
+			I = overlay_image(icon, "water_attachment")
+	if(I)
+		AddOverlays(I)
+
+/obj/item/sampler/attack(atom/target, mob/user, proximity_flag, click_parameters)
+	if(!vial || vial.reagents.total_volume)
+		return ..()
+	else
+		return FALSE
+
+/obj/item/sampler/afterattack(atom/target, mob/user, proximity_flag, click_parameters)
+	if(!vial || vial.reagents.total_volume)
+		return ..()
+	switch(attachment)
+		if(SAMPLE_BIO)
+			if(istype(target, /mob/living/simple_animal))
+				var/mob/living/simple_animal/fauna = target
+				if(!fauna.has_cells)
+					return
+				vial.reagents.add_reagent(/singleton/reagent/cells, 10, fauna.sample_data)
+				. = TRUE
+			if(istype(target, /obj/structure/flora))
+				var/obj/structure/flora/flora = target
+				if(flora.is_rock)
+					return
+				vial.reagents.add_reagent(/singleton/reagent/cells, 10, flora.sample_data)
+				. = TRUE
+		if(SAMPLE_SOIL)
+			if(istype(target, /obj/structure/flora))
+				var/obj/structure/flora/rock = target
+				if(!rock.is_rock)
+					return
```

Comment:
this applies to all of these branches in this proc
there should be some feedback about the sampler not being able to make sense of what you are trying to sample
like a buzz-beep, or it could have a tiny screen displaying a message or something

Reply:
or something like

```
to_chat(user, SPAN_NOTICE("\The [src]'s attachment lets out a sharp buzz, as it tries to sample \the [target]"))
```